SafetyFirst/Set.fs

```diff
@@ -0,0 +1,220 @@
+    if Seq.isEmpty xs
+    then Error <| SeqIsEmpty "Assertion that a sequence is not empty failed."
+    else Ok (NonEmpty (Set.ofSeq xs))
+
+  /// <summary>
+  /// Asserts that <c>xs</c> is not empty, creating a NonEmptySet.
+  /// Returns a SeqIsEmpty Error if <c>xs</c> is empty.
+  /// </summary>
+  let inline ofSeq' xs = ofSeqSafe xs 
+
+  /// <summary>
+  /// Builds a NonEmptySet that contains the same elements as the given NonEmptySeq.
+  /// </summary>
+  let ofNonEmptySeq (NonEmpty xs : NonEmptySeq<_>) : NonEmptySet<_> = 
+    NonEmpty (Set.ofSeq xs)
+
+  /// <summary>
+  /// Asserts that <c>xs</c> is not empty, creating a NonEmptySet.
+  /// Returns a SeqIsEmpty Error if <c>xs</c> is empty.
+  /// </summary>
+  let ofListSafe xs : Result<NonEmptySet<_>, _> = 
+    if List.isEmpty xs
+    then Error <| SeqIsEmpty "Assertion that a sequence is not empty failed."
+    else Ok (NonEmpty (Set.ofList xs))
+
+  /// <summary>
+  /// Asserts that <c>xs</c> is not empty, creating a NonEmptySet.
+  /// Returns a SeqIsEmpty Error if <c>xs</c> is empty.
+  /// </summary>
+  let inline ofList' xs = ofListSafe xs 
+
+  /// <summary>
+  /// Builds a NonEmptySet that contains the same elements as the given NonEmptyList.
+  /// </summary>
+  let ofNonEmptyList (NonEmpty xs : NonEmptyList<_>) : NonEmptySet<_> = 
+    NonEmpty (Set.ofList xs)
+
+  /// <summary>
+  /// Splits the set into two sets containing the elements for which the given predicate returns true and false respectively.
+  /// </summary>
+  let partition f (NonEmpty xs : NonEmptySet<_>) = Set.partition f xs
+
+  /// <summary>
+  /// Returns a new set with the given element removed. 
+  /// No exception is raised if the set doesn't contain the given element.
+  /// </summary>
+  let remove element (NonEmpty xs : NonEmptySet<_>) = Set.remove element xs
+
+  /// <summary>
+  /// Builds an array that contains the elements of the set in order.
+  /// </summary>
+  let toArray (NonEmpty xs : NonEmptySet<_>) = Set.toArray xs
+
+  /// <summary>
+  /// Builds a NonEmptyArray that contains the elements of the set in order.
+  /// </summary>
+  let toNonEmptyArray (NonEmpty xs : NonEmptySet<_>) : NonEmptyArray<_> = NonEmpty (Set.toArray xs)
+
+  /// <summary>
+  /// Builds a list that contains the elements of the set in order.
+  /// </summary>
+  let toList (NonEmpty xs : NonEmptySet<_>) = Set.toList xs
+
+  /// <summary>
+  /// Builds a NonEmptyList that contains the elements of the set in order.
+  /// </summary>
+  let toNonEmptyList (NonEmpty xs : NonEmptySet<_>) : NonEmptyList<_> = NonEmpty (Set.toList xs)
+
+  /// <summary>
+  /// Builds a seq that contains the elements of the set in order.
+  /// </summary>
+  let toSeq (NonEmpty xs : NonEmptySet<_>) = Set.toSeq xs
+
+  /// <summary>
+  /// Builds a NonEmptySeq that contains the elements of the set in order.
+  /// </summary>
+  let toNonEmptySeq (NonEmpty xs : NonEmptySet<_>) : NonEmptySeq<_> = NonEmpty (Set.toSeq xs)

```

Comment:
can this just return `upcast xs`?  That would be more performant

Reply:
I can turn it into: `NonEmpty (upcast xs)`, but I can't figure out how to do better than that.  I'm guessing you need covariance to be able to do it.